### PR TITLE
Mail validation

### PR DESCRIPTION
**Pull request by snimavat**
https://github.com/9ci/domain9/issues/2708

### CODE REVIEW

File: rally-domain/src/main/groovy/yakworks/rally/mail/EmailService.groovy [hunk 1]
```diff
@@ -94,6 +97,25 @@ abstract class EmailService {
 
         /** Specify Reply-To address */
         String replyTo;
+    }
 
+    /**
+     * Verify if given email is valid
+     * Accepts single or comma seperated list of emails
+     */
+    Result validateEmail(String mail) {
```
Comment (basejump) [inline rally-domain/src/main/groovy/yakworks/rally/mail/EmailService.groovy:106]: Use EmailUtils.validateEmail

**Review by basejump — CHANGES_REQUESTED** [review]

File: rally-domain/src/main/groovy/yakworks/rally/mail/EmailService.groovy (hunk 1 above)
Comment (snimavat) [reply rally-domain/src/main/groovy/yakworks/rally/mail/EmailService.groovy]: Done

File: rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy [hunk 2]
```diff
@@ -38,8 +40,11 @@ class MailMessageSender {
         Result result
         try {
             mailTo = convertMailMessage(mailMessage)
-            // mailService.send should never throw ex and should return result
-            result = emailService.send(mailTo)
+            boolean valid = validateEmail(mailTo.to[0]) //this will throw a problem when validation fails.
+            if(valid) {
```
Comment (basejump) [inline rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy:44]: this is kind of confusing right? what happens when its false or when would it be false?

File: rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy [hunk 3]
```diff
@@ -38,8 +40,11 @@ class MailMessageSender {
...
+                // mailService.send should never throw ex and should return result
+                result = emailService.send(mailTo)
+            }
         } catch(ex){
             //in convertMailMessage might throw an ex if the attachmentId is bad or not found
             result = Problem.of(ex)
```
Comment (basejump) [inline rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy:50]: if validateEmail throws then we dont want to rewrap it in a problem do we?

**Review by basejump — CHANGES_REQUESTED** [review]

File: rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy (hunk 2 above)
Comment (snimavat) [reply rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy]: Fixed

File: rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy (hunk 3 above)
Comment (snimavat) [reply rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy:53]: Fixed

File: rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy [hunk 4]
```diff
@@ -24,21 +24,27 @@ class EmailUtils {
...
+                } catch (AddressException e) {
+                    throw Problem.of("validation.problem")
+                        .payload(addr)
+                        .detail("Invalid email address [$addr], " + e.message).toException()
+                }
             }
-            isValid = true;
         } catch (AddressException e) {
-            throw Problem.of("validation.problem").detail(e.message).toException()
+            throw Problem.of("validation.problem").payload(emails).detail(e.message).toException()
```
Comment (basejump) [inline rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy:46]: what this one for? the `InternetAddress.parse(emails)`?

File: rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy [hunk 5]
```diff
@@ -33,25 +36,31 @@ class MailMessageSender {
...
-        MailTo mailTo
+    Result send(MailMessage mailMessage) {
         Result result
+
         try {
-            mailTo = convertMailMessage(mailMessage)
-            // mailService.send should never throw ex and should return result
+            MailTo mailTo = convertMailMessage(mailMessage)
+            //validation throws a problem exception if validation fails
+            validateEmail(mailTo.to[0])
```
Comment (basejump) [inline rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy:45]: why are you only validating the first one?

File: rally-domain/src/main/groovy/yakworks/rally/mail/mailgun/MailgunService.groovy [hunk 6]
```diff
@@ -66,7 +66,7 @@ class MailgunService extends EmailService {
             log.debug("Mail gun response : domain:$domain, id:${resp.id}, message:${resp.message}")
             return Result.OK().payload([id: resp.id, message: resp.message])
         } catch(FeignException e){
-            log.error("Mailgun Send Failure", e)
+            log.error("Mailgun failed to send email to ${mailMsg.to}", e)
```
Comment (basejump) [inline rally-domain/src/main/groovy/yakworks/rally/mail/mailgun/MailgunService.groovy:69]: log more so we can debug it, goal is to diagnose the 500's right?, and we should log what we set to `Message message` right, thats whats being sent over. the way this is not wont help us sort anything out that I can see.

File: rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy [hunk 7]
```diff
@@ -24,21 +24,27 @@ class EmailUtils {
...
+        }
         try {
             InternetAddress[] addys = InternetAddress.parse(emails)
-            addys.each{
-                it.validate()
+            for (InternetAddress addr : addys) {
+                try {
+                    addr.validate()
+                } catch (AddressException e) {
+                    throw Problem.of("validation.problem")
```
Comment (basejump) [inline rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy:40]: image

has a specific meaning. expects name to be passed. or we can consider using DataProblem.
Also below, comment on why your adding payload reference to the InternetAddress.

**Review by basejump — CHANGES_REQUESTED** [review]
see comments

File: rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy (hunk 4 above)
Comment (snimavat) [reply rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy:46]: Yes, parse can throw error too

File: rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy (hunk 5 above)
Comment (snimavat) [reply rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy:45]: `convertMailMessage` puts just single value in [mailTo](https://github.com/yakworks/gorm-tools/blob/a0ce39d0218e78d28f4353b66bb968ac94c3bc97/rally-domain/src/main/groovy/yakworks/rally/mail/services/MailMessageSender.groovy#L89)

File: rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy (hunk 7 above)
Comment (snimavat) [reply rally-domain/src/main/groovy/yakworks/rally/mail/EmailUtils.groovy]: Fixed payload, changed to use DataProblem

**Pull request merged**